client/state/data-layer/wpcom/account-recovery/lookup/index.js

```diff
@@ -35,20 +35,20 @@ export const fromApi = data => ( [
 	},
 ] );
 
-export const onSuccess = ( { dispatch }, action, next, data ) => {
-	if ( validate( data ) ) {
-		dispatch( {
-			type: ACCOUNT_RECOVERY_RESET_OPTIONS_RECEIVE,
-			items: fromApi( data ),
-		} );
-	} else {
+export const storeReceivedOptions = ( { dispatch }, action, next, data ) => {
+	if ( ! validate( data ) ) {
 		throw Error( 'Unexpected response format' );
 	}
+
+	dispatch( {
+		type: ACCOUNT_RECOVERY_RESET_OPTIONS_RECEIVE,
+		items: fromApi( data ),
```

Comment:
I don't have any problem with how you are doing this here but I also thought up an idea using [lodash.tap()](https://lodash.com/docs/#tap) which I thought you might be interested in seeing

```js
const validate = data => {
	if ( data.age < 0 ) {
		throw new RangeError( 'Age must be positive' );
	}
}

…

items: fromApi( tap( data, validate ) ),
```

the main difference here is that we can see in one swoop where the validation happens and the validation is explicitly a side-effecting operation _only_.